### ISSUE TITLE
auto register_tb

### ISSUE THREAD

**Issue by foez-ahmed** [Chip-DSi/simple_processor_BUET#96]
Testbench Needed

**Comment by anindyakchoudhury** [comment]
![image](https://github.com/Chip-DSi/simple_processor_BUET/assets/69748292/7f74e88a-f757-4735-a6f4-2e9d91d98837)
Bhaiya, have I handled the matter of negedge arst_ni correctly?

**Comment by anindyakchoudhury** [comment]
` // monitor and check
  task static start_checking();
    fork
      forever begin
        @(posedge clk_i);
        if (q_o === ref_data) begin
          pass++;
        end else begin
          fail++;
          //$display("ERROR: Expected q_o = %h, Got q_o = %h at time %t", expected_q, q_o, $time);
        end

        if (~arst_ni) begin
          ref_data = RESET_VALUE;
        end else if (en_i) begin
          ref_data <= d_i;
        end
      end
    join_none
  endtask`

**Comment by anindyakchoudhury** [comment]
```verilog
task static start_checking();
fork
forever begin
@(posedge clk_i or negedge arst_ni);
if (~arst_ni) begin
if (q_o != RESET_VALUE) begin
fail++;
$display("ERROR: Expected q_o = %h, Got q_o = %h at time %t due to reset", RESET_VALUE, q_o, $time);
end
end else begin
if (q_o == expected_q) begin
pass++;
end else begin
fail++;
$display("ERROR: Expected q_o = %h, Got q_o = %h at time %t", expected_q, q_o, $time);
end
if (en_i) begin
expected_q <= d_i;
end
end
end
join_none
endtask
```

**Comment by foez-ahmed** [comment]
Auto-closing issue

**Comment by foez-ahmed** [comment]
I'm not quite sure why need a separate RTL for a register. I'm pretty sure you don't need a register with non-zero reset value. Let's dicsuss this tomorrow (12 July) in call at 11:30 AM?

**Comment by anindyakchoudhury** [comment]
@foez-ahmed Bhaiya,

1. we could have used it from the system Verilog submodule. But after running the code, so many errors came in—for example, the default param package issue and a bunch of others.
2. So, I reverted our make file to the original one we used. Then I copied the necessary files from the system verilog, edited as we needed it (like we only need Num_Rs to be 2).
3. We are almost done, bhaiya. Hopefully, we will complete the testbench of the top module by tonight and push it.
4. Up until 6 pm tomorrow, I have a pre-scheduled program to attend. May we hold a meeting at 7.45 pm tomorrow? By then, I will be home, and the project will be completed with every part.
